routing/bicycle_directions.cpp

```diff
@@ -222,39 +165,172 @@ Index::FeaturesLoaderGuard & BicycleDirectionsEngine::GetLoader(MwmSet::MwmId co
 {
-  pathSegment.Clear();
-
-  if (!uniNodeId.GetFeature().IsValid())
-  {
-    ASSERT(false, ());
+  if (!featureId.IsValid())
     return;
-  }
 
   FeatureType ft;
-  if (!GetLoader(uniNodeId.GetFeature().m_mwmId)
-           .GetFeatureByIndex(uniNodeId.GetFeature().m_index, ft))
-  {
-    // The feature can't be read, therefore path geometry can't be
-    // loaded.
-    return;
-  }
+  GetLoader(featureId.m_mwmId).GetFeatureByIndex(featureId.m_index, ft);
 
   auto const highwayClass = ftypes::GetHighwayClass(ft);
   ASSERT_NOT_EQUAL(highwayClass, ftypes::HighwayClass::Error, ());
   ASSERT_NOT_EQUAL(highwayClass, ftypes::HighwayClass::Undefined, ());
 
   pathSegment.m_highwayClass = highwayClass;
   pathSegment.m_isLink = ftypes::IsLinkChecker::Instance()(ft);
-
   ft.GetName(FeatureType::DEFAULT_LANG, pathSegment.m_name);
-
-  pathSegment.m_nodeId = uniNodeId;
   pathSegment.m_onRoundabout = ftypes::IsRoundAboutChecker::Instance()(ft);
-  pathSegment.m_path = path;
-  // @TODO(bykoianko) It's better to fill pathSegment.m_weight.
+}
+
+void BicycleDirectionsEngine::GetUniNodeIdAndAdjacentEdges(
+    IRoadGraph::TEdgeVector const & outgoingEdges, FeatureID const & inFeatureId,
+    uint32_t startSegId, uint32_t endSegId, bool inIsForward, UniNodeId & uniNodeId,
+    BicycleDirectionsEngine::AdjacentEdges & adjacentEdges)
+{
+  // Outgoing edge angle is not used for bicycle routing.
+  adjacentEdges.m_outgoingTurns.isCandidatesAngleValid = false;
+  adjacentEdges.m_outgoingTurns.candidates.reserve(outgoingEdges.size());
+  uniNodeId = UniNodeId(inFeatureId, startSegId, endSegId, inIsForward);
+
+  for (auto const & edge : outgoingEdges)
+  {
+    auto const & outFeatureId = edge.GetFeatureId();
+    // Checking if |edge| is a fake edge.
+    if (!outFeatureId.IsValid())
+      continue;
+
+    FeatureType ft;
+    GetLoader(outFeatureId.m_mwmId).GetFeatureByIndex(outFeatureId.m_index, ft);
+
+    auto const highwayClass = ftypes::GetHighwayClass(ft);
+    ASSERT_NOT_EQUAL(highwayClass, ftypes::HighwayClass::Error, ());
+    ASSERT_NOT_EQUAL(highwayClass, ftypes::HighwayClass::Undefined, ());
+    adjacentEdges.m_outgoingTurns.candidates.emplace_back(0.0 /* angle */, uniNodeId, highwayClass);
+  }
+}
+
+bool BicycleDirectionsEngine::IsJoint(IRoadGraph::TEdgeVector const & ingoingEdges,
+                                      IRoadGraph::TEdgeVector const & outgoingEdges,
+                                      Edge const & ingoingRouteEdge, Edge const & outgoingRouteEdge)
+{
+  if (ingoingRouteEdge.GetFeatureId() != outgoingRouteEdge.GetFeatureId())
+    return true;
+
+  FeatureID const & featureId = ingoingRouteEdge.GetFeatureId();
+  uint32_t const segOut = outgoingRouteEdge.GetSegId();
+  for (Edge const & e : ingoingEdges)
+  {
+    // It's necessary to compare segments for cases when |featureId| is a loop.
+    if (e.GetFeatureId() != featureId || abs(static_cast<int>(segOut - e.GetSegId())) != 1)
```

Comment:
I'm adding the following comment to the code:
```
  // When feature id is changed at a junction this junction should be considered as a joint.
  //
  // If a feature id is not changed at a junction but the junction has some ingoing or outgoing edges with
  // different feature ids, the junction should be considered as a joint.
  //
  // If a feature id is not changed at a junction and all ingoing and outgoing edges of the junction has
  // the same feature id, the junction still may be considered as a joint.
  // It happens in case of self intersected features. For example:
  //            *--Seg3--*
  //            |        |
  //          Seg4      Seg2
  //            |        |
  //   *--Seg0--*--Seg1--*
  // The common point of segments 0, 1 and 4 should be considered as a joint.
```